// @alchemy/aa-accounts/dist/cjs/src/msca/account-loupe/decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.accountLoupeDecorators = void 0;
const IAccountLoupe_js_1 = require("../abis/IAccountLoupe.js");
const accountLoupeDecorators = (account) => ({
    getExecutionFunctionConfig: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupe_js_1.IAccountLoupeAbi,
        functionName: "getExecutionFunctionConfig",
        args: [selector],
    }),
    getExecutionHooks: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupe_js_1.IAccountLoupeAbi,
        functionName: "getExecutionHooks",
        args: [selector],
    }),
    getPreValidationHooks: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupe_js_1.IAccountLoupeAbi,
        functionName: "getPreValidationHooks",
        args: [selector],
    }),
    getInstalledPlugins: async () => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupe_js_1.IAccountLoupeAbi,
        functionName: "getInstalledPlugins",
    }),
});
exports.accountLoupeDecorators = accountLoupeDecorators;
