// @alchemy/aa-accounts/dist/cjs/src/light-account/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLightAccountProvider = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const account_js_1 = require("./account.js");
const schema_js_1 = require("./schema.js");
const utils_js_1 = require("./utils.js");
const createLightAccountProvider = (config_) => {
    const config = schema_js_1.LightAccountProviderConfigSchema.parse(config_);
    return new aa_core_1.SmartAccountProvider({
        rpcProvider: config.rpcProvider,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        opts: config.opts,
    }).connect((rpcClient) => new account_js_1.LightSmartContractAccount({
        rpcClient,
        initCode: config.initCode,
        owner: config.owner,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        factoryAddress: config.factoryAddress ??
            (0, utils_js_1.getDefaultLightAccountFactoryAddress)(config.chain, config.version),
        accountAddress: config.accountAddress,
    }));
};
exports.createLightAccountProvider = createLightAccountProvider;
