// @alchemy/aa-accounts/dist/cjs/src/light-account/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightAccountUnsupported1271Factories = exports.LightAccountUnsupported1271Impls = exports.getDefaultLightAccountFactoryAddress = exports.LightAccountVersions = void 0;
const chains_1 = require("viem/chains");
exports.LightAccountVersions = {
    "v1.0.1": {
        factoryAddress: "0x000000893A26168158fbeaDD9335Be5bC96592E2".toLowerCase(),
        implAddress: "0xc1b2fc4197c9187853243e6e4eb5a4af8879a1c0".toLowerCase(),
    },
    "v1.0.2": {
        factoryAddress: "0x00000055C0b4fA41dde26A74435ff03692292FBD".toLowerCase(),
        implAddress: "0x5467b1947F47d0646704EB801E075e72aeAe8113".toLowerCase(),
    },
    "v1.1.0": {
        factoryAddress: "0x00004EC70002a32400f8ae005A26081065620D20".toLowerCase(),
        implAddress: "0xae8c656ad28F2B59a196AB61815C16A0AE1c3cba".toLowerCase(),
    },
};
const getDefaultLightAccountFactoryAddress = (chain, version = "v1.1.0") => {
    switch (chain.id) {
        case chains_1.mainnet.id:
        case chains_1.sepolia.id:
        case chains_1.goerli.id:
        case chains_1.polygon.id:
        case chains_1.polygonMumbai.id:
        case chains_1.optimism.id:
        case chains_1.optimismGoerli.id:
        case chains_1.optimismSepolia.id:
        case chains_1.arbitrum.id:
        case chains_1.arbitrumGoerli.id:
        case chains_1.arbitrumSepolia.id:
        case chains_1.base.id:
        case chains_1.baseGoerli.id:
        case chains_1.baseSepolia.id:
            return exports.LightAccountVersions[version].factoryAddress;
    }
    throw new Error(`no default light account factory contract exists for ${chain.name}`);
};
exports.getDefaultLightAccountFactoryAddress = getDefaultLightAccountFactoryAddress;
exports.LightAccountUnsupported1271Impls = [
    exports.LightAccountVersions["v1.0.1"],
    exports.LightAccountVersions["v1.0.2"],
];
exports.LightAccountUnsupported1271Factories = new Set(exports.LightAccountUnsupported1271Impls.map((x) => x.factoryAddress));
