// @0xpolygonid/js-sdk/dist/node/cjs/iden3comm/handlers/fetch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchHandler = void 0;
const constants_1 = require("../constants");
const constants_2 = require("../constants");
const verifiable_1 = require("../../verifiable");
const credentials_1 = require("../../credentials");
const utils_1 = require("../../utils");
const js_jwz_1 = require("@iden3/js-jwz");
const js_iden3_core_1 = require("@iden3/js-iden3-core");
const uuid = __importStar(require("uuid"));
/**
 *
 * Allows to handle Credential offer protocol message and return fetched credential
 *
 * @public

 * @class FetchHandler
 * @implements implements IFetchHandler interface
 */
class FetchHandler {
    /**
     * Creates an instance of FetchHandler.
     * @param {IPackageManager} _packerMgr - package manager to unpack message envelope
     */
    constructor(_packerMgr, opts) {
        this._packerMgr = _packerMgr;
        this.opts = opts;
    }
    /**
     * Handles only messages with credentials/1.0/offer type
     *
     * @param {
     *     offer: Uint8Array; offer - raw offer message
     *     opts
     *   }) options how to fetch credential
     * @returns `Promise<W3CCredential[]>`
     */
    async handleCredentialOffer(offer, opts) {
        if (!opts) {
            opts = {
                mediaType: constants_1.MediaType.ZKPMessage
            };
        }
        if (opts.mediaType === constants_1.MediaType.SignedMessage && !opts.packerOptions) {
            throw new Error(`jws packer options are required for ${constants_1.MediaType.SignedMessage}`);
        }
        const offerMessage = await FetchHandler.unpackMessage(this._packerMgr, offer, constants_2.PROTOCOL_MESSAGE_TYPE.CREDENTIAL_OFFER_MESSAGE_TYPE);
        const credentials = [];
        for (let index = 0; index < offerMessage.body.credentials.length; index++) {
            const credentialInfo = offerMessage.body.credentials[index];
            const guid = uuid.v4();
            const fetchRequest = {
                id: guid,
                typ: opts.mediaType,
                type: constants_2.PROTOCOL_MESSAGE_TYPE.CREDENTIAL_FETCH_REQUEST_MESSAGE_TYPE,
                thid: offerMessage.thid ?? guid,
                body: {
                    id: credentialInfo.id
                },
                from: offerMessage.to,
                to: offerMessage.from
            };
            const msgBytes = utils_1.byteEncoder.encode(JSON.stringify(fetchRequest));
            const packerOpts = opts.mediaType === constants_1.MediaType.SignedMessage
                ? opts.packerOptions
                : {
                    provingMethodAlg: js_jwz_1.proving.provingMethodGroth16AuthV2Instance.methodAlg
                };
            const senderDID = js_iden3_core_1.DID.parse(offerMessage.to);
            const token = utils_1.byteDecoder.decode(await this._packerMgr.pack(opts.mediaType, msgBytes, {
                senderDID,
                ...packerOpts
            }));
            let message;
            try {
                if (!offerMessage?.body?.url) {
                    throw new Error(`could not fetch W3C credential, body url is missing`);
                }
                const resp = await fetch(offerMessage.body.url, {
                    method: 'post',
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded',
                        ...(opts.headers ?? {})
                    },
                    body: token
                });
                if (resp.status !== 200) {
                    throw new Error(`could not fetch W3C credential, ${credentialInfo?.id}`);
                }
                message = await resp.json();
                credentials.push(verifiable_1.W3CCredential.fromJSON(message.body.credential));
            }
            catch (e) {
                throw new Error(`could not fetch W3C credential, ${credentialInfo?.id}, error: ${e.message ?? e}`);
            }
        }
        return credentials;
    }
    /**
     * @inheritdoc IFetchHandler#handleCredentialFetchRequest
     */
    async handleCredentialFetchRequest(envelope) {
        const msgRequest = await FetchHandler.unpackMessage(this._packerMgr, envelope, constants_2.PROTOCOL_MESSAGE_TYPE.CREDENTIAL_FETCH_REQUEST_MESSAGE_TYPE);
        if (!msgRequest.to) {
            throw new Error("failed request. empty 'to' field");
        }
        if (!msgRequest.from) {
            throw new Error("failed request. empty 'from' field");
        }
        const issuerDID = js_iden3_core_1.DID.parse(msgRequest.to);
        const userDID = js_iden3_core_1.DID.parse(msgRequest.from);
        const credId = msgRequest.body?.id;
        if (!credId) {
            throw new Error('nvalid credential id in fetch request body');
        }
        if (!this.opts?.credentialWallet) {
            throw new Error('please, provide credential wallet in options');
        }
        const cred = await this.opts.credentialWallet.findById(credId);
        if (!cred) {
            throw new Error('credential not found');
        }
        const userToVerifyDID = (0, credentials_1.getUserDIDFromCredential)(issuerDID, cred);
        if (userToVerifyDID.string() !== userDID.string()) {
            throw new Error('credential subject is not a sender DID');
        }
        return this._packerMgr.pack(constants_1.MediaType.PlainMessage, utils_1.byteEncoder.encode(JSON.stringify({
            id: uuid.v4(),
            type: constants_2.PROTOCOL_MESSAGE_TYPE.CREDENTIAL_ISSUANCE_RESPONSE_MESSAGE_TYPE,
            typ: msgRequest.typ ?? constants_1.MediaType.PlainMessage,
            thid: msgRequest.thid ?? uuid.v4(),
            body: { credential: cred },
            from: msgRequest.to,
            to: msgRequest.from
        })), {});
    }
    /**
     * @inheritdoc IFetchHandler#handleIssuanceResponseMessage
     */
    async handleIssuanceResponseMessage(envelop) {
        const issuanceMsg = await FetchHandler.unpackMessage(this._packerMgr, envelop, constants_2.PROTOCOL_MESSAGE_TYPE.CREDENTIAL_ISSUANCE_RESPONSE_MESSAGE_TYPE);
        if (!this.opts?.credentialWallet) {
            throw new Error('please provide credential wallet in options');
        }
        if (!issuanceMsg.body?.credential) {
            throw new Error('credential is missing in issuance response message');
        }
        await this.opts.credentialWallet.save(verifiable_1.W3CCredential.fromJSON(issuanceMsg.body.credential));
        return Promise.resolve(Uint8Array.from([]));
    }
    /**
     * @inheritdoc IFetchHandler#unpackMessage
     */
    static async unpackMessage(packerMgr, envelope, messageType) {
        const { unpackedMessage: message } = await packerMgr.unpack(envelope);
        const msgRequest = message;
        if (message.type !== messageType) {
            throw new Error('Invalid message type');
        }
        return msgRequest;
    }
}
exports.FetchHandler = FetchHandler;
