// @0xpolygonid/js-sdk/dist/node/cjs/circuits/linked-nullifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedNullifierPubSignals = exports.LinkedNullifierInputs = void 0;
const js_iden3_core_1 = require("@iden3/js-iden3-core");
const utils_1 = require("../utils");
/**
 * LinkedNullifier circuit representation
 * Inputs and public signals declaration, marshalling and parsing
 *
 * @beta
 * @class LinkedNullifierInputs
 */
class LinkedNullifierInputs {
    // InputsMarshal returns Circom private inputs for nullifier.circom
    inputsMarshal() {
        const s = {
            linkNonce: this.linkNonce.toString(),
            issuerClaim: this.issuerClaim.marshalJson(),
            userGenesisID: this.id.bigInt().toString(),
            claimSubjectProfileNonce: this.claimSubjectProfileNonce.toString(),
            claimSchema: this.issuerClaim.getSchemaHash().bigInt().toString(),
            verifierID: this.verifierID?.bigInt().toString() ?? '0',
            nullifierSessionID: this.nullifierSessionID.toString()
        };
        return utils_1.byteEncoder.encode(JSON.stringify(s));
    }
}
exports.LinkedNullifierInputs = LinkedNullifierInputs;
// LinkedNullifierPubSignals nullifier.circom public signals
/**
 * public signals
 *
 * @beta
 * @class LinkedNullifierPubSignals
 */
class LinkedNullifierPubSignals {
    /**
     * PubSignalsUnmarshal unmarshal nullifier.circom public inputs to LinkedNullifierPubSignals
     *
     * @beta
     * @param {Uint8Array} data
     * @returns LinkedNullifierPubSignals
     */
    pubSignalsUnmarshal(data) {
        const len = 4;
        const sVals = JSON.parse(utils_1.byteDecoder.decode(data));
        if (sVals.length !== len) {
            throw new Error(`invalid number of Output values expected ${len} got ${sVals.length}`);
        }
        this.nullifier = BigInt(sVals[0]);
        this.linkID = BigInt(sVals[1]);
        this.verifierID = js_iden3_core_1.Id.fromBigInt(BigInt(sVals[2]));
        this.nullifierSessionID = BigInt(sVals[3]);
        return this;
    }
}
exports.LinkedNullifierPubSignals = LinkedNullifierPubSignals;
