// @alchemy/aa-accounts/dist/cjs/src/msca/plugins/session-key/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendedSessionKeyPlugin = void 0;
const plugin_js_1 = require("./plugin.js");
const utils_js_1 = require("./utils.js");
const ExtendedSessionKeyPlugin_ = {
    ...plugin_js_1.SessionKeyPlugin,
    accountMethods: (account) => ({
        isAccountSessionKey: async ({ key, pluginAddress, }) => {
            const contract = plugin_js_1.SessionKeyPlugin.getContract(account.rpcProvider, pluginAddress);
            return await contract.read.isSessionKeyOf([
                await account.getAddress(),
                key,
            ]);
        },
        getAccountSessionKeys: async ({ pluginAddress, }) => {
            const contract = plugin_js_1.SessionKeyPlugin.getContract(account.rpcProvider, pluginAddress);
            return await contract.read.sessionKeysOf([await account.getAddress()]);
        },
    }),
    providerMethods: (provider) => {
        const generated = plugin_js_1.SessionKeyPlugin.providerMethods(provider);
        return {
            ...generated,
            removeAccountSessionKey: async ({ key }, overrides) => {
                const sessionKeysToRemove = await (0, utils_js_1.buildSessionKeysToRemoveStruct)(provider, [key]);
                return generated.removeSessionKey({ args: [key, sessionKeysToRemove[0].predecessor] }, overrides);
            },
        };
    },
};
exports.ExtendedSessionKeyPlugin = ExtendedSessionKeyPlugin_;
