// @alchemy/aa-accounts/dist/cjs/src/msca/plugins/session-key/signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionKeySigner = exports.SESSION_KEY_SIGNER_TYPE_PFX = exports.createSessionKeySignerSchema = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const accounts_1 = require("viem/accounts");
const zod_1 = require("zod");
const createSessionKeySignerSchema = () => {
    return zod_1.z.object({
        storageType: zod_1.z
            .union([zod_1.z.literal("local-storage"), zod_1.z.literal("session-storage")])
            .default("local-storage"),
        storageKey: zod_1.z.string().default("session-key-signer:session-key"),
        fallbackSigner: zod_1.z.custom((signer) => aa_core_1.SignerSchema.parse(signer)),
    });
};
exports.createSessionKeySignerSchema = createSessionKeySignerSchema;
exports.SESSION_KEY_SIGNER_TYPE_PFX = "alchemy:session-key";
class SessionKeySigner {
    constructor(config_) {
        Object.defineProperty(this, "signerType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "inner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "keyActive", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "fallback", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storageType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storageKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "getAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                if (!this.keyActive) {
                    return this.fallback.getAddress();
                }
                return this.inner.getAddress();
            }
        });
        Object.defineProperty(this, "signMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (msg) => {
                if (!this.keyActive) {
                    return this.fallback.signMessage(msg);
                }
                return this.inner.signMessage(msg);
            }
        });
        Object.defineProperty(this, "signTypedData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (params) => {
                if (!this.keyActive) {
                    return this.fallback.signTypedData(params);
                }
                return this.inner.signTypedData(params);
            }
        });
        Object.defineProperty(this, "isKeyActive", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                return this.keyActive;
            }
        });
        Object.defineProperty(this, "setKeyActive", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (active) => {
                return (this.keyActive = active);
            }
        });
        Object.defineProperty(this, "generateNewKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                const storage = this.storageType === "session-storage" ? sessionStorage : localStorage;
                const newKey = (0, accounts_1.generatePrivateKey)();
                storage.setItem(this.storageKey, newKey);
                this.inner = aa_core_1.LocalAccountSigner.privateKeyToAccountSigner(newKey);
                this.keyActive = true;
                return this.inner.inner.address;
            }
        });
        const config = (0, exports.createSessionKeySignerSchema)().parse(config_);
        this.signerType = `${exports.SESSION_KEY_SIGNER_TYPE_PFX}:${config.fallbackSigner.signerType}`;
        this.storageKey = config.storageKey;
        this.storageType = config.storageType;
        const sessionKey = (() => {
            const storage = config.storageType === "session-storage"
                ? sessionStorage
                : localStorage;
            const key = storage.getItem(this.storageKey);
            if (key) {
                return key;
            }
            else {
                const newKey = (0, accounts_1.generatePrivateKey)();
                storage.setItem(this.storageKey, newKey);
                return newKey;
            }
        })();
        this.inner = aa_core_1.LocalAccountSigner.privateKeyToAccountSigner(sessionKey);
        this.keyActive = true;
        this.fallback = config.fallbackSigner;
    }
}
exports.SessionKeySigner = SessionKeySigner;
