// @alchemy/aa-accounts/dist/cjs/src/msca/plugin-manager/installPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeInstallPluginUserOperation = exports.installPlugin = void 0;
const viem_1 = require("viem");
const IPlugin_js_1 = require("../abis/IPlugin.js");
const IPluginManager_js_1 = require("../abis/IPluginManager.js");
async function installPlugin(provider, params, overrides) {
    const callData = await encodeInstallPluginUserOperation(provider, params);
    return provider.sendUserOperation(callData, overrides);
}
exports.installPlugin = installPlugin;
async function encodeInstallPluginUserOperation(provider, params) {
    const pluginManifest = await provider.rpcClient.readContract({
        abi: IPlugin_js_1.IPluginAbi,
        address: params.pluginAddress,
        functionName: "pluginManifest",
    });
    const manifestHash = params.manifestHash ??
        (0, viem_1.keccak256)((0, viem_1.encodeFunctionResult)({
            abi: IPlugin_js_1.IPluginAbi,
            functionName: "pluginManifest",
            result: pluginManifest,
        }));
    return (0, viem_1.encodeFunctionData)({
        abi: IPluginManager_js_1.IPluginManagerAbi,
        functionName: "installPlugin",
        args: [
            params.pluginAddress,
            manifestHash,
            params.pluginInitData ?? "0x",
            params.dependencies ?? [],
        ],
    });
}
exports.encodeInstallPluginUserOperation = encodeInstallPluginUserOperation;
