// @alchemy/aa-accounts/dist/cjs/src/nani-account/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNaniAccountProvider = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const account_js_1 = require("./account.js");
const schema_js_1 = require("./schema.js");
const utils_js_1 = require("./utils.js");
const createNaniAccountProvider = (config_) => {
    const config = schema_js_1.NaniAccountProviderConfigSchema.parse(config_);
    return new aa_core_1.SmartAccountProvider({
        rpcProvider: config.rpcProvider,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        opts: config.opts,
    }).connect((rpcClient) => new account_js_1.NaniAccount({
        rpcClient,
        initCode: config.initCode,
        owner: config.owner,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        factoryAddress: config.factoryAddress ??
            (0, utils_js_1.getDefaultNaniAccountFactoryAddress)(config.chain),
        accountAddress: config.accountAddress,
        salt: config.salt,
        index: config.index,
    }));
};
exports.createNaniAccountProvider = createNaniAccountProvider;
