// @alchemy/aa-accounts/dist/cjs/src/msca/builder/wrapped-signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrapWith712SignerMethods = void 0;
const viem_1 = require("viem");
const plugin_js_1 = require("../plugins/multi-owner/plugin.js");
const WrapWith712SignerMethods = (acct) => {
    const owner = acct.getOwner();
    if (!owner) {
        throw new Error("owner is required for use with signer methods");
    }
    const signWith712Wrapper = async (msg) => {
        const [, name, version, chainId, , salt] = await acct.rpcProvider.readContract({
            abi: plugin_js_1.MultiOwnerPluginAbi,
            address: plugin_js_1.MultiOwnerPlugin.meta.addresses[acct.rpcProvider.chain.id],
            functionName: "eip712Domain",
            account: await acct.getAddress(),
        });
        return owner.signTypedData({
            domain: {
                chainId: Number(chainId),
                name,
                salt,
                verifyingContract: await acct.getAddress(),
                version,
            },
            types: {
                AlchemyModularAccountMessage: [{ name: "message", type: "bytes" }],
            },
            message: {
                message: msg,
            },
            primaryType: "AlchemyModularAccountMessage",
        });
    };
    return {
        getDummySignature() {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        },
        signUserOperationHash(uoHash) {
            return owner.signMessage((0, viem_1.hexToBytes)(uoHash));
        },
        signMessage(msg) {
            return signWith712Wrapper((0, viem_1.hashMessage)(typeof msg === "string" && !(0, viem_1.isBytes)(msg)
                ? msg
                : {
                    raw: msg,
                }));
        },
        signTypedData(params) {
            return signWith712Wrapper((0, viem_1.hashTypedData)(params));
        },
    };
};
exports.WrapWith712SignerMethods = WrapWith712SignerMethods;
