// @alchemy/aa-accounts/dist/esm/src/msca/builder/index.js
import { BaseSmartContractAccount, createBaseSmartAccountParamsSchema, isSigner, } from "@alchemy/aa-core";
import {} from "viem";
import { z } from "zod";
import { pluginManagerDecorator } from "../plugin-manager/decorator.js";
const zCompleteBuilder = z.object({
    executor: z.custom(),
    signer: z.custom(),
    factory: z.custom(),
});
export const ModularAccountBuilderParamsSchema = () => createBaseSmartAccountParamsSchema().extend({
    owner: z.custom(isSigner),
    index: z.bigint().optional(),
});
export class MSCABuilder {
    constructor() {
        Object.defineProperty(this, "executor", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "factory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
    withExecutor(executor) {
        return Object.assign(this, { executor });
    }
    withSigner(methods) {
        return Object.assign(this, { signer: methods });
    }
    withFactory(initCode) {
        return Object.assign(this, { factory: initCode });
    }
    build(params) {
        const builder = this;
        const { signer, executor, factory } = zCompleteBuilder.parse(builder);
        return new (class DynamicMSCA extends BaseSmartContractAccount {
            constructor() {
                super(...arguments);
                Object.defineProperty(this, "providerDecorators_", {
                    enumerable: true,
                    configurable: true,
                    writable: true,
                    value: [pluginManagerDecorator]
                });
                Object.defineProperty(this, "providerDecorators", {
                    enumerable: true,
                    configurable: true,
                    writable: true,
                    value: (p) => {
                        if (!p.isConnected() && p.account !== this) {
                            throw new Error("provider should be connected if it is being decorated by the account");
                        }
                        return this.providerDecorators_.reduce((acc, decorator) => ({
                            ...acc,
                            ...decorator(p),
                        }), {});
                    }
                });
                Object.defineProperty(this, "extendWithPluginMethods", {
                    enumerable: true,
                    configurable: true,
                    writable: true,
                    value: (plugin) => {
                        const methods = plugin.accountMethods(this);
                        const result = Object.assign(this, methods);
                        result.providerDecorators_.push(plugin.providerMethods);
                        return result;
                    }
                });
                Object.defineProperty(this, "addProviderDecorator", {
                    enumerable: true,
                    configurable: true,
                    writable: true,
                    value: (decorator) => {
                        this.providerDecorators_.push(decorator);
                        return this;
                    }
                });
            }
            getDummySignature() {
                return signer(this).getDummySignature();
            }
            encodeExecute(target, value, data) {
                return executor(this).encodeExecute(target, value, data);
            }
            encodeBatchExecute(txs) {
                return executor(this).encodeBatchExecute(txs);
            }
            signMessage(msg) {
                return signer(this).signMessage(msg);
            }
            signTypedData(params) {
                return signer(this).signTypedData(params);
            }
            signUserOperationHash(uoHash) {
                return signer(this).signUserOperationHash(uoHash);
            }
            getAccountInitCode() {
                return factory(this);
            }
        })(params);
    }
}
