// @0xpolygonid/js-sdk/dist/node/esm/storage/blockchain/state.js
import { Contract, JsonRpcProvider } from 'ethers';
import { StateTransitionPubSignals } from '../../circuits';
import { byteEncoder } from '../../utils';
import abi from './abi/State.json';
export /** @type {EthConnectionConfig} - default configuration for EthConnectionConfig */ const defaultEthConnectionConfig = {
    url: 'http://localhost:8545',
    defaultGasLimit: 600000,
    minGasPrice: '0',
    maxGasPrice: '100000000000',
    confirmationBlockCount: 5,
    confirmationTimeout: 600000,
    contractAddress: '',
    receiptTimeout: 600000,
    rpcResponseTimeout: 5000,
    waitReceiptCycleTime: 30000,
    waitBlockCycleTime: 3000
};
/**
 *
 *
 * @public
 * @class EthStateStorage
 * @implements implements IStateStorage interface
 */
export class EthStateStorage {
    /**
     * Creates an instance of EthStateStorage.
     * @param {EthConnectionConfig} [ethConfig=defaultEthConnectionConfig]
     */
    constructor(ethConfig = defaultEthConnectionConfig) {
        this.ethConfig = ethConfig;
        this.provider = new JsonRpcProvider(this.ethConfig.url);
        this.stateContract = new Contract(this.ethConfig.contractAddress, abi, this.provider);
    }
    /** {@inheritdoc IStateStorage.getLatestStateById} */
    async getLatestStateById(id) {
        const rawData = await this.stateContract.getStateInfoById(id);
        const stateInfo = {
            id: BigInt(rawData[0]),
            state: BigInt(rawData[1]),
            replacedByState: BigInt(rawData[2]),
            createdAtTimestamp: BigInt(rawData[3]),
            replacedAtTimestamp: BigInt(rawData[4]),
            createdAtBlock: BigInt(rawData[5]),
            replacedAtBlock: BigInt(rawData[6])
        };
        return stateInfo;
    }
    /** {@inheritdoc IStateStorage.publishState} */
    async publishState(proof, signer) {
        const contract = this.stateContract.connect(signer);
        const stateTransitionPubSig = new StateTransitionPubSignals();
        stateTransitionPubSig.pubSignalsUnmarshal(byteEncoder.encode(JSON.stringify(proof.pub_signals)));
        const { userId, oldUserState, newUserState, isOldStateGenesis } = stateTransitionPubSig;
        const payload = [
            userId.bigInt().toString(),
            oldUserState.bigInt().toString(),
            newUserState.bigInt().toString(),
            isOldStateGenesis,
            proof.proof.pi_a.slice(0, 2),
            [
                [proof.proof.pi_b[0][1], proof.proof.pi_b[0][0]],
                [proof.proof.pi_b[1][1], proof.proof.pi_b[1][0]]
            ],
            proof.proof.pi_c.slice(0, 2)
        ];
        const feeData = await this.provider.getFeeData();
        const maxFeePerGas = defaultEthConnectionConfig.maxFeePerGas
            ? BigInt(defaultEthConnectionConfig.maxFeePerGas)
            : feeData.maxFeePerGas;
        const maxPriorityFeePerGas = defaultEthConnectionConfig.maxPriorityFeePerGas
            ? BigInt(defaultEthConnectionConfig.maxPriorityFeePerGas)
            : feeData.maxPriorityFeePerGas;
        const gasLimit = await contract.transitState.estimateGas(...payload);
        const txData = await contract.transitState.populateTransaction(...payload);
        const request = {
            to: txData.to,
            data: txData.data,
            gasLimit,
            maxFeePerGas,
            maxPriorityFeePerGas
        };
        const tx = await signer.sendTransaction(request);
        const txnReceipt = await tx.wait();
        if (!txnReceipt) {
            throw new Error(`transaction: ${tx.hash} failed to mined`);
        }
        const status = txnReceipt.status;
        const txnHash = txnReceipt.hash;
        if (!status) {
            throw new Error(`transaction: ${txnHash} failed to mined`);
        }
        return txnHash;
    }
    /** {@inheritdoc IStateStorage.getGISTProof} */
    async getGISTProof(id) {
        const data = await this.stateContract.getGISTProof(id);
        return {
            root: BigInt(data.root.toString()),
            existence: data.existence,
            siblings: data.siblings?.map((sibling) => BigInt(sibling.toString())),
            index: BigInt(data.index.toString()),
            value: BigInt(data.value.toString()),
            auxExistence: data.auxExistence,
            auxIndex: BigInt(data.auxIndex.toString()),
            auxValue: BigInt(data.auxValue.toString())
        };
    }
    /** {@inheritdoc IStateStorage.getGISTRootInfo} */
    async getGISTRootInfo(id) {
        const data = await this.stateContract.getGISTRootInfo(id);
        return {
            root: BigInt(data.root.toString()),
            replacedByRoot: BigInt(data.replacedByRoot.toString()),
            createdAtTimestamp: BigInt(data.createdAtTimestamp.toString()),
            replacedAtTimestamp: BigInt(data.replacedAtTimestamp.toString()),
            createdAtBlock: BigInt(data.createdAtBlock.toString()),
            replacedAtBlock: BigInt(data.replacedAtBlock.toString())
        };
    }
}
