// @alchemy/aa-accounts/dist/esm/src/msca/account-loupe/decorator.js
import { IAccountLoupeAbi } from "../abis/IAccountLoupe.js";
export const accountLoupeDecorators = (account) => ({
    getExecutionFunctionConfig: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupeAbi,
        functionName: "getExecutionFunctionConfig",
        args: [selector],
    }),
    getExecutionHooks: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupeAbi,
        functionName: "getExecutionHooks",
        args: [selector],
    }),
    getPreValidationHooks: async (selector) => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupeAbi,
        functionName: "getPreValidationHooks",
        args: [selector],
    }),
    getInstalledPlugins: async () => account.rpcProvider.readContract({
        address: await account.getAddress(),
        abi: IAccountLoupeAbi,
        functionName: "getInstalledPlugins",
    }),
});
